Service IP Allocator doesn't handle certain IPv6 ranges

### What happened?

We ran into an issue with an old version of k8s 1.18 that we were running where subnet for cluster IP is given as `fc::/112`. 

The control plane comes up just fine but then when you try to create a k8s service, it fails due to a broken behavior with the IP Allocator 

```bash
# kubectl apply -f /etc/kubernetes/etcd-service.yaml
The Service "etcd" is invalid: spec.clusterIP: Invalid value: "?fc00000000000000000000000077a5": must be empty, 'None', or a valid IP address
```
Since this is not really a supported version of the k8s, I went around digging up a bit more to see what is the regular behavior and found this to be very inconsistent across a few release branches. 

> All of the example below are the output of the test case `TestAllocateNextFC` referenced below. 

### On master Branch
```bash
k8s/kubernetes/pkg/registry/core/service/ipallocator/ipallocator_test.go:82: invalid address [252 0 0 0 0 0 0 0 0 0 0 0 0 1 1]
--- FAIL: TestAllocateNextFC (0.00s)
FAIL
FAIL	k8s.io/kubernetes/pkg/registry/core/service/ipallocator	2.040s
```

### On `release-1.27`
```bash
k8s/kubernetes/pkg/registry/core/service/ipallocator/ipallocator_test.go:82: invalid address [252 0 0 0 0 0 0 0 0 0 0 0 0 1 1]
--- FAIL: TestAllocateNextFC (0.00s)
FAIL
FAIL	k8s.io/kubernetes/pkg/registry/core/service/ipallocator	1.599s
```

### on `release-1.26`
```bash
=== RUN   TestAllocateNextFC
fc::2c23
--- PASS: TestAllocateNextFC (0.00s)
PASS
ok  	k8s.io/kubernetes/pkg/registry/core/service/ipallocator	1.533s
```

### On `release-1.24`
```bash
go test -timeout 30m -run ^TestAllocateNextFC$ k8s.io/kubernetes/pkg/registry/core/service/ipallocator -v

=== RUN   TestAllocateNextFC
fc::a06e
--- PASS: TestAllocateNextFC (0.00s)
PASS
ok  	k8s.io/kubernetes/pkg/registry/core/service/ipallocator	0.566s
```

### On `release-1.18`
```bash
```bash
go test -timeout 30m -run ^TestAllocateNextFC$ k8s.io/kubernetes/pkg/registry/core/service/ipallocator -v

=== RUN   TestAllocateNextFC
?fc00000000000000000000000077a5
--- PASS: TestAllocateNextFC (0.00s)
PASS
ok  	k8s.io/kubernetes/pkg/registry/core/service/ipallocator	0.566s
```

### What did you expect to happen?

IP Allocator should return a valid IP address or return an error if the CIDR ranges are invalid. 

### How can we reproduce it (as minimally and precisely as possible)?

```go
func TestAllocateNextFC(t *testing.T) {
	_, cidr, err := netutils.ParseCIDRSloppy("fc::/112")
	if err != nil {
		t.Fatal(err)
	}
        // Need to be replaced with the suitable handler for that version of the branch
	r, err := newTestAllocator(cidr)
	if err != nil {
		t.Fatal(err)
	}
	defer r.Destroy()
	ip, err := r.AllocateNext()
	if err != nil {
		t.Fatalf("%s", err.Error())
	}
	fmt.Println(ip.String())
}
```
Add the above Test case under the IP allocator test case 

### Anything else we need to know?

_No response_

### Kubernetes version

<details>

NA

</details>


### Cloud provider

<details>
NA
</details>


### OS version

<details>
NA
</details>


### Install tools

<details>
NA
</details>


### Container runtime (CRI) and version (if applicable)

<details>
NA
</details>


### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>
NA
</details>

--- a/pkg/registry/core/service/ipallocator/ipallocator.go
+++ b/pkg/registry/core/service/ipallocator/ipallocator.go
@@ -484,11 +484,21 @@ func (dry dryRunAllocator) EnableMetrics() {
 // TODO: move it to k8s.io/utils/net, this is the same as current AddIPOffset()
 // but using netip.Addr instead of net.IP
 func addOffsetAddress(address netip.Addr, offset uint64) (netip.Addr, error) {
-	addressBig := big.NewInt(0).SetBytes(address.AsSlice())
-	r := big.NewInt(0).Add(addressBig, big.NewInt(int64(offset)))
-	addr, ok := netip.AddrFromSlice(r.Bytes())
+	addressBytes := address.AsSlice()
+	addressBig := big.NewInt(0).SetBytes(addressBytes)
+	r := big.NewInt(0).Add(addressBig, big.NewInt(int64(offset))).Bytes()
+	// r must be 4 or 16 bytes depending of the ip family
+	// bigInt conversion to bytes will not take this into consideration
+	// and drop the leading zeros, so we have to take this into account.
+	lenDiff := len(addressBytes) - len(r)
+	if lenDiff > 0 {
+		r = append(make([]byte, lenDiff), r...)
+	} else if lenDiff < 0 {
+		return netip.Addr{}, fmt.Errorf("invalid address %v", r)
+	}
+	addr, ok := netip.AddrFromSlice(r)
 	if !ok {
-		return netip.Addr{}, fmt.Errorf("invalid address %v", r.Bytes())
+		return netip.Addr{}, fmt.Errorf("invalid address %v", r)
 	}
 	return addr, nil
 }

--- a/pkg/registry/core/service/ipallocator/ipallocator_test.go
+++ b/pkg/registry/core/service/ipallocator/ipallocator_test.go
@@ -588,6 +588,18 @@ func Test_addOffsetAddress(t *testing.T) {
 			want:    netip.MustParseAddr("192.168.0.128"),
 		},
 		{
+			name:    "IPv4 with leading zeros",
+			address: netip.MustParseAddr("0.0.1.8"),
+			offset:  138,
+			want:    netip.MustParseAddr("0.0.1.146"),
+		},
+		{
+			name:    "IPv6 with leading zeros",
+			address: netip.MustParseAddr("00fc::1"),
+			offset:  255,
+			want:    netip.MustParseAddr("fc::100"),
+		},
+		{
 			name:    "IPv6 offset 255",
 			address: netip.MustParseAddr("2001:db8:1::101"),
 			offset:  255,
@@ -642,6 +654,16 @@ func Test_broadcastAddress(t *testing.T) {
 			name:   "ipv6",
 			subnet: netip.MustParsePrefix("fd00:1:2:3::/64"),
 			want:   netip.MustParseAddr("fd00:1:2:3:FFFF:FFFF:FFFF:FFFF"),
+		},
+		{
+			name:   "ipv6 00fc::/112",
+			subnet: netip.MustParsePrefix("00fc::/112"),
+			want:   netip.MustParseAddr("fc::ffff"),
+		},
+		{
+			name:   "ipv6 fc00::/112",
+			subnet: netip.MustParsePrefix("fc00::/112"),
+			want:   netip.MustParseAddr("fc00::ffff"),
 		},
 	}
 	for _, tt := range tests {
@@ -886,6 +908,25 @@ func Test_ipIterator_Number(t *testing.T) {
 			}
 		})
 	}
+}
+
+func TestAllocateNextFC(t *testing.T) {
+	_, cidr, err := netutils.ParseCIDRSloppy("fc::/112")
+	if err != nil {
+		t.Fatal(err)
+	}
+	t.Logf("CIDR %s", cidr)
+
+	r, err := newTestAllocator(cidr)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer r.Destroy()
+	ip, err := r.AllocateNext()
+	if err != nil {
+		t.Fatalf("wrong ip %s : %v", ip, err)
+	}
+	t.Log(ip.String())
 }
 
 func BenchmarkIPAllocatorAllocateNextIPv4Size1048574(b *testing.B) {

